larpandoracontent/LArTwoDReco/LArClusterAssociation/CrossGapsExtensionAlgorithm.cc

```diff
@@ -51,72 +51,90 @@ void CrossGapsExtensionAlgorithm::GetListOfCleanClusters(const ClusterList *cons
+
+            // Get closest pair of vertices from pointing clusters
+            LArPointingCluster::Vertex closestVertex1, closestVertex2;
+
+            try
+            {
+                LArPointingClusterHelper::GetClosestVertices(pointingCluster1, pointingCluster2, closestVertex1, closestVertex2);
+            }
+            catch (StatusCodeException &)
+            {
                 continue;
+            }
 
-            const Cluster *const pClusterInner(pointingClusterInner.GetCluster());
-            const Cluster *const pClusterOuter(pointingClusterOuter.GetCluster());
+            // Check that these vertices have associated proximity and pointing information
+            if (!this->IsAssociated(closestVertex1, closestVertex2))
+                continue;
 
-            const float lengthSquaredInner(LArClusterHelper::GetLengthSquared(pClusterInner));
-            const float lengthSquaredOuter(LArClusterHelper::GetLengthSquared(pClusterOuter));
+            // Check that these vertices lie across a registered gap
+            if (!(LArGeometryHelper::IsInGap(this->GetPandora(), closestVertex1.GetPosition(), closestVertex2.GetPosition(),
+                hitType1, m_maxGapTolerance)))
+                continue;
 
-            (void) clusterAssociationMatrix[pClusterInner].insert(ClusterAssociationMap::value_type(pClusterOuter,
-                ClusterAssociation(ClusterAssociation::INNER, ClusterAssociation::OUTER, ClusterAssociation::STRONG, lengthSquaredOuter)));
-            (void) clusterAssociationMatrix[pClusterOuter].insert(ClusterAssociationMap::value_type(pClusterInner,
-                ClusterAssociation(ClusterAssociation::OUTER, ClusterAssociation::INNER, ClusterAssociation::STRONG, lengthSquaredInner)));
-        }
-    }
-}
+            // Calculate figure of merit for this pair of clusters and require a positive answer
+            const float clusterLength1(LArClusterHelper::GetLength(pCluster1));
+            const float clusterLength2(LArClusterHelper::GetLength(pCluster2));
+            const float clusterSeparation((closestVertex1.GetPosition() - closestVertex2.GetPosition()).GetMagnitude());
 
-//------------------------------------------------------------------------------------------------------------------------------------------
+            const float cosTheta(-1.f * closestVertex1.GetDirection().GetDotProduct(closestVertex2.GetDirection()));
+            const float logTheta(-1.f * std::log10(1.0 - cosTheta + std::numeric_limits<float>::epsilon()));
 
-void CrossGapsExtensionAlgorithm::BuildPointingClusterList(const ClusterVector &clusterVector, LArPointingClusterList &innerPointingClusterList,
-    LArPointingClusterList &outerPointingClusterList) const
-{
-    // Convert each input cluster into a pointing cluster
-    LArPointingClusterList pointingClusterList;
+            const float figureOfMerit12(logTheta * (2.f * clusterLength2 - clusterSeparation));
```

Comment:
Figure of merit might warrant a ATTN comment to (briefly) explain aims.

Reply:
Yes, that's a good point - I've added some words of explanation. 

As a TODO for the future, we ought to think about how we analyse length and pointing information when stitching together tracks. By default, we've been making the joins that yield the longest tracks, but this isn't always the best decision.